### PR TITLE
Start RPC server before loading snapshot

### LINKED ISSUE #1800: Start the RPC server as soon as possible (before loading snapshots)
**Issue summary**

The Forest daemon initiates the RPC server after loading snapshot. This is less than ideal since loading a snapshot can take quite a while and we would like the RPC server up-and-running as soon as possible. The code should be refactored such that the RPC server is booted as soon as it can be.

**Other information and links**

### PR DESCRIPTION
**Pull request by connormullett**
**Summary of changes**
Changes introduced in this pull request:
- Change the ordering of tasks in daemon.rs to start the RPC server prior to loading the snapshot

**Reference issue to close (if applicable)**

Closes #1800 

**Other information and links**

### CODE REVIEW

File: forest/src/daemon.rs [hunk 1]
```diff
@@ -339,6 +294,56 @@ pub(super) async fn start(config: Config) {
...
+        };
+        set_proofs_parameter_cache_dir_env(&config.client.data_dir);
+
+        get_params_default(&config.client.data_dir, SectorSizeOpt::Keys, false)
+            .await
+            .unwrap();
+    }
+
+    // Start services
+    let p2p_task = task::spawn(async {
```
Comment (LesnyRumcajs) [inline forest/src/daemon.rs:343]: An edge case, but IMO worth considering. What will happen if we call some of the commands like `forest net peers` while the snapshot is loading? The p2p service was initialized before RPC, right now I'm not entirely sure what will happen and do we have a nice error handling.

Comment (lemmih) [reply forest/src/daemon.rs:344]: Good question. Right `forest net peers` blocks until the p2p service is up and running. I feel this is alright but I haven't given it much thought.

**Comment by lemmih** [comment]
I modified the code to properly shutdown if ctrl-c is detected.

File: forest/src/daemon.rs (hunk 1 above)
Comment (LesnyRumcajs) [reply forest/src/daemon.rs]: I'd say an immediate answer that the Forest service is busy may be more convenient for whoever is integrating with it, but it's fine for now, we should revisit those calls later on when covering them with a nice suite of tests.

File: forest/src/daemon.rs [hunk 2]
```diff
@@ -298,48 +265,101 @@ pub(super) async fn start(config: Config) {
...
             .unwrap_or_else(|_| cli_error_and_die("could not bind to {rpc_address}", 1));
 
-        Some(task::spawn(async move {
+        let rpc_state_manager = Arc::clone(&state_manager);
+        let rpc_chain_store = Arc::clone(&chain_store);
+
+        services.push(task::spawn(async move {
             info!("JSON-RPC endpoint started at {}", config.client.rpc_address);
-            start_rpc::<_, _, FullVerifier, cns::FullConsensus>(
+            // XXX: The JSON error message are a nightmare to print.
```
Comment (LesnyRumcajs) [inline forest/src/daemon.rs:289]: I'm not sure what does this comment refers to? do we need it?

**Review by LesnyRumcajs — APPROVED** [review]

File: forest/src/daemon.rs (hunk 2 above)
Comment (lemmih) [reply forest/src/daemon.rs:289]: The line below silently ignores a json-rpc error. Ideally I would like to log it but there's no standard way of pretty-printing those errors.

Comment (lemmih) [reply forest/src/daemon.rs:289]: (The error was also silently ignored previously.)

**Review by lemmih — APPROVED** [review]

**Pull request merged**